libraries/AP_ToshibaCAN/AP_ToshibaCAN.cpp

```diff
@@ -0,0 +1,287 @@
+        */
+
+        // return semaphore
+        _rc_out_sem.give();
+
+        // send unlock command
+        if (send_stage == 0) {
+            timeout = uavcan::MonotonicTime::fromUSec(AP_HAL::micros64() + loop_interval_us);
+            if (!write_frame(unlock_frame, timeout)) {
+                _fail2++;
+                continue;
+            }
+            send_stage = 1;
+        }
+
+        // wait for space in buffer to send output to motor bank1
+        if (send_stage == 1) {
+            timeout = uavcan::MonotonicTime::fromUSec(AP_HAL::micros64() + loop_interval_us);
+            if (!write_frame(mot_rot_frame1, timeout)) {
+                _fail3++;
+                continue;
+            }
+        }
+
+        // success!
+        send_stage = 0;
+
+        // record success so we don't send this frame again
+        last_send_us = last_update_us;
+
+        /*
+        // wait for space in buffer to send output to motor bank2
+        timeout = uavcan::MonotonicTime::fromUSec(AP_HAL::micros64() + loop_interval_us);
+        if (!write_frame(mot_rot_frame2, timeout)) {
+            _fail4++;
+            continue;
+        }
+
+        // wait for space in buffer to send output to motor bank3
+        timeout = uavcan::MonotonicTime::fromUSec(AP_HAL::micros64() + loop_interval_us);
+        if (!write_frame(mot_rot_frame3, timeout)) {
+            _fail5++;
+            continue;
+        }*/
+
+        _success++;
+    }
+}
+
+// write frame on CAN bus
+bool AP_ToshibaCAN::write_frame(uavcan::CanFrame &out_frame, uavcan::MonotonicTime timeout)
+{
+    // wait for space in buffer to send unlock command
+    uavcan::CanSelectMasks inout_mask;
+    inout_mask.write = 1 << CAN_IFACE_INDEX;
+    _select_frames[CAN_IFACE_INDEX] = &out_frame;
+    _can_driver->select(inout_mask, _select_frames, timeout);
+
+    // double check space to send is available
+    if (!inout_mask.write) {
+        return false;
+    }
+
+    // send frame
+    int8_t res = _can_driver->getIface(CAN_IFACE_INDEX)->send(out_frame, timeout, 0);
+
+    // return results (1 == success)
+    return res == 1;
+}
+
+// called from SRV_Channels
+void AP_ToshibaCAN::update()
+{
+    if (_rc_out_sem.take(1)) {
+        uint16_t pwm_out[16] {};
+        hal.rcout->read(pwm_out, 16);
+        for (uint8_t i = 0; i < MIN(TOSHIBACAN_MAX_NUM_ESCS, 16); i++) {
+            if (pwm_out[i] <= 1000) {
+                _scaled_output[i] = 0;
+            } else if (pwm_out[i] >= 2000) {
+                _scaled_output[i] = TOSHIBACAN_OUTPUT_MAX;
+            } else {
+                _scaled_output[i] = TOSHIBACAN_OUTPUT_MIN + (pwm_out[i] - 1000) * 0.001f * (TOSHIBACAN_OUTPUT_MAX - TOSHIBACAN_OUTPUT_MIN);
+            }
+        }
+        last_update_us = AP_HAL::micros64();
+        _rc_out_sem.give();
+    } else {
+        debug_can(2, "ToshibaCAN: failed to get PWM semaphore on write\n\r");
+    }
+
+    // debug
+    static uint16_t counter = 0;
```

Comment:
i think you probably meant to remove this debug code? Or #if with a DEBUG macro?